### ISSUE TITLE
Type id not exposed for `JsonTypeInfo.As.EXTERNAL_PROPERTY` even when `visible` set to `true`

### ISSUE THREAD

**Issue by libetl** [FasterXML/jackson-databind#1288]
Jackson : 2.8.0.rc2

Hello,

I have these following jsons to deserialize, with a type property ("form_of_payment") located outside a polymorphic class ("payment_details") :
```json
    {"form_of_payment":"INDIVIDUAL_CREDIT_CARD","payment_details":{"card_holder_first_name":"John", "card_holder_last_name":"Doe",  "number":"XXXXXXXXXXXXXXXX", "expiryDate":"MM/YY"}
```
```json
    {"form_of_payment":"INSTRUMENTED_CREDIT_CARD","payment_details":{"payment_instrument_id":"00000000-0000-0000-0000-000000000000", "name":"Mr John Doe encrypted credit card"}
```

I have to use the `@JsonTypeInfo` annotation on the PaymentDetails class, to inform Jackson that there are several kinds of PaymentDetails.

The include must be a `include = JsonTypeInfo.As.EXTERNAL_PROPERTY`.

Once the TypeResolver has done his job, the JsonTypeInfo is removed from the feed, which annoys me a lot (the "form_of_payment" also must be read for business purposes). If it is an `EXTERNAL_PROPERTY`, there is no way to restore it (it works with `PROPERTY`, but the JsonTypeInfo must be inside the same object)

Can you add a spare visibility property to the `@JsonTypeInfo` to find it in the parent object ?
Here is a link to where it happens :  https://github.com/FasterXML/jackson-databind/blob/master/src/main/java/com/fasterxml/jackson/databind/jsontype/impl/AsArrayTypeDeserializer.java#L99

Thanks.

(I have seen the https://github.com/FasterXML/jackson-databind/issues/408 issue BTW)

**Comment by cowtowncoder** [comment]
Did you check the javadocs? If so, wouldn't `visible`:

http://fasterxml.github.io/jackson-annotations/javadoc/2.7/com/fasterxml/jackson/annotation/JsonTypeInfo.html#visible()

do what you ask?

**Comment by libetl** [comment]
visible does not work with external_property

**Comment by cowtowncoder** [comment]
@libetl issue #408 claims it does, so I'd need a unit test to prove otherwise to see when and how it does not work as expected, reproducible against either 2.8.0 or 2.7.5 (earlier branches are closed for development).

**Comment by libetl** [comment]
with pleasure. Give me some time to try that.

**Comment by libetl** [comment]
Please have a look at this failing test (I have cloned HEAD today) :
https://gist.github.com/libetl/9c6ec1820444370c3c55670058f57426

**Comment by cowtowncoder** [comment]
@libetl thank you for reproduction. I hope to have a look at it soon.

**Comment by libetl** [comment]
I found out that this issue is closely linked to the builder serialization.
I have updated my gist, with two versions of the same unmarshalling.
The first one (without builder) works, the second one fails.

**Comment by libetl** [comment]
NOW, I know.
The Jira issue JACKSON-831 change on the BeanDeserializer was not applied on the BuilderBasedDeserializer (which appears to be a copy/paste).
Do you want me to submit a PR ?

**Comment by cowtowncoder** [comment]
Note: regressed in 3.0.0, re-fixed in 3.0.4/3.1.0.
